### PR TITLE
[model] Change the semantics for setting the batchsize

### LINKED ISSUE #299: Overlap with minibatch and inputshape
Input shape takes batch dimension as input. However, this can be in conflict with `minibatch` argument in ini file format.

### PR DESCRIPTION
**Pull request by kparichay**
Batchsize as a property was earlier set in input_shape for a layer
as well as for the model. However, this was conflicting and prone to errors.

This patch disables setting batch size while setting the input_shape.
If batch_size is more than 1 in the input_shape, a warning is issued.
And this value is later overriden by the batchsize set for the model.
This same batchsize is also used for the dataset as well.

Resolves #299

**Self evaluation:**
1. Build test: [x]Passed [ ]Failed [ ]Skipped
2. Run test: [x]Passed [ ]Failed [ ]Skipped

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

File: nntrainer/src/neuralnet.cpp [hunk 1]
```diff
@@ -536,6 +529,9 @@ sharedTensor NeuralNetwork::forwarding(sharedTensor input) {
 sharedTensor NeuralNetwork::forwarding(sharedTensor input, sharedTensor label) {
   sharedTensor X;
 
+  if (input->getDim().batch() > batch_size)
```
Comment (zhoonit) [inline nntrainer/src/neuralnet.cpp:532]: Q. is it valid for `input->getDim().batch() < batch_size` ?

File: nntrainer/src/neuralnet.cpp [hunk 2]
```diff
@@ -536,6 +529,9 @@ sharedTensor NeuralNetwork::forwarding(sharedTensor input) {
...
+    throw std::logic_error("Error: mismatch in batchsize for data and model.");
```
Comment (zhoonit) [inline nntrainer/src/neuralnet.cpp:533]: Could you add `std::logic_error` to `nntrainer_exception_boundary`?

Comment (kparichay) [reply nntrainer/src/neuralnet.cpp:533]: done

File: nntrainer/src/parse_util.cpp [hunk 3]
```diff
@@ -380,10 +384,10 @@ unsigned int parseDataProperty(std::string property) {
   return (unsigned int)DataBuffer::PropertyType::unknown;
 }
 
-int setInt(int &val, std::string str) {
+int setUint(unsigned int &val, const std::string &str) {
   int status = ML_ERROR_NONE;
   try {
-    val = std::stoi(str.c_str());
+    val = (unsigned int)std::stoi(str.c_str());
```
Comment (zhoonit) [inline nntrainer/src/parse_util.cpp:390]:
```suggestion
    val = std::stoul(str.c_str());
```

File: nntrainer/src/tensor_dim.cpp [hunk 4]
```diff
@@ -44,11 +44,21 @@ void TensorDim::resetLen() {
...
+    throw std::invalid_argument(
+      "[TensorDim] Tensor Dimension index should be between 0 and 4");
+
+  return dim[idx];
+}
+
 void TensorDim::setTensorDim(unsigned int idx, unsigned int value) {
-  if (value == 0) {
+  if (idx < 0 || idx > MAXDIM)
+    throw std::invalid_argument(
```
Comment (zhoonit) [inline nntrainer/src/tensor_dim.cpp:57]: how about `throw std::out_of_range` ? We might need to differentiate those two cases.

File: nntrainer/src/tensor_dim.cpp [hunk 5]
```diff
@@ -61,7 +71,7 @@ int TensorDim::setTensorDim(std::string input_shape) {
     std::sregex_iterator(input_shape.begin(), input_shape.end(), words_regex);
   auto words_end = std::sregex_iterator();
   int cur_dim = std::distance(words_begin, words_end);
-  if (cur_dim > MAXDIM) {
+  if (cur_dim <= 0 || cur_dim > MAXDIM) {
     ml_loge("Tensor Dimension should be less than 4");
```
Comment (zhoonit) [inline nntrainer/src/tensor_dim.cpp:75]:
```suggestion
    ml_loge("Tensor Dimension should be between 1 and 4");
```

Comment (zhoonit) [reply nntrainer/src/tensor_dim.cpp:75]: Slightly out of topic though, I think line 80~83 / 85~86 can be safely deleted.

File: nntrainer/src/layer.cpp [hunk 6]
```diff
@@ -158,8 +158,29 @@ void Layer::setProperty(const PropertyType type, const std::string &value) {
     break;
   case PropertyType::input_shape:
     if (!value.empty()) {
+      unsigned int cache_batch_size = 1;
+      /** cache original value of batch size */
+      if (input_dim.batch()) {
```
Comment (kparichay) [inline nntrainer/src/layer.cpp:163]: @zhoonit Updated.

File: nntrainer/src/parse_util.cpp (hunk 3 above)
Comment (kparichay) [reply nntrainer/src/parse_util.cpp:390]: done.

File: nntrainer/src/tensor_dim.cpp (hunk 4 above)
Comment (kparichay) [reply nntrainer/src/tensor_dim.cpp:57]: done.

File: nntrainer/src/tensor_dim.cpp (hunk 5 above)
Comment (kparichay) [reply nntrainer/src/tensor_dim.cpp:75]: Done.
Removed both checks and update `stoi` to `stoul` on line 80.

**Pull request merged**